runtime/Cpp/runtime/src/atn/ArrayPredictionContext.cpp

```diff
@@ -10,49 +10,61 @@
 
 using namespace antlr4::atn;
 
+PredictionContextItem::PredictionContextItem(const Ref<PredictionContext>& p, size_t s)
+  : parent(p),
+    returnState(s) { }
+
+bool PredictionContextItem::operator == (const PredictionContextItem& o) const {
+    return returnState == o.returnState && *parent == *o.parent;
+}
+
 ArrayPredictionContext::ArrayPredictionContext(Ref<SingletonPredictionContext> const& a)
-  : ArrayPredictionContext({ a->parent }, { a->returnState }) {
+  : PredictionContext(a->cachedHashCode),
+    contexts(1, PredictionContextItem(a->parent, a->returnState)) {
+   assert(a);
 }
 
-ArrayPredictionContext::ArrayPredictionContext(std::vector<Ref<PredictionContext>> const& parents_,
-                                               std::vector<size_t> const& returnStates)
-  : PredictionContext(calculateHashCode(parents_, returnStates)), parents(parents_), returnStates(returnStates) {
-    assert(parents.size() > 0);
-    assert(returnStates.size() > 0);
+ArrayPredictionContext::ArrayPredictionContext(std::vector<PredictionContextItem> contexts_)
+  : PredictionContext(calculateHashCode(contexts_)),
+    contexts(std::move(contexts_)) {
```

Comment:
A bit unusual to pass in a copy and move that, instead of using a reference and copy that. Any good reason why you didn't stay with the `const&` approach used throught the runtime?

Reply:
The intention here is to avoid a call to the vector copy constructor (which will copy the contents). There is a related move in PredictionContext.cpp,

Ref<ArrayPredictionContext> M = std::make_shared<ArrayPredictionContext>(std::move(contexts));

Reply:
You are making a copy, when you call the function. I don't understand what your goal is here. And I don't see any need to change the established approach. By using `const&` in the function signature you avoid a copy, but later on you have to copy anyway (to keep a local duplicate). I don't see a reason why we should change that.

Reply:
After adding the proper rvalue reference semantics this doesn't hit the vector copy constructor here, which was the initial intention. Unfortunately, I didn't quite understand all the move semantics stuff in my first attempt.